rfcs/007 - Alternative Ways to Create Mappers.md

```diff
@@ -0,0 +1,122 @@
+SELECT iasql_install('https://github.com/someModuleDeveloper/awsEc2Module');
+```
+That's where we started, but it might be useful to assess different paths we might take for creation of an IaSQL module/mapper.
+
+## Proposal
+
+To create a module, we need a mapper. I'll continue this RFC with my proposed solution for mappers, since a module is a set of mappers, and entities (with a bit of oversimplification).
+
+First, let's see what we need to create a mapper:
+- A cloud id function
+- An equality checker function
+- An entity, that the mapper manages it
+- Cloud CRUD functions
+- Ability to call other mappers' CRUD functions
+
+We need to get the mapper-like behavior working. I'll explain some ways we might consider for that purpose.
+
+- Remote Mappers
+  - Mappers maintained inside IaSQL infrastructure
+  - Mappers maintained outside IaSQL infrastructure
+- In-IaSQL-Container Mappers
+  - Mapper Logic Inside IaSQL Engine
+  - Mapper Logic Inside Postgres
+
+### A Stand-Alone Mapper Design (Remote Mappers)
+
+We can run standalone versions of mappers. A standalone mapper means that its logic won't be inside the Postgres, nor in the engine code. But it'll run anywhere outside these two places. Therefore, we need a communication mechanism for this design (which can be something like gRPC, or HTTP as in other places). We also might want to run that standalone mapper on our own server, or on somewhere else. Two approaches:
+
+1. Standalone mappers are served on the IaSQL server: the mapper is submitted to our system, and we'll run it in a container. Then our engine will communicate with that mapper through a communication mechanism. This reduces latency, since the calls will be intra-network, but adds the burden of maintaining the standalone mappers to our side.
+![In-house Standalone Mapper](./assets/008/Mapper-1.png)
+2. Standalone mappers served on servers other than IaSQL: the mapper providers can host that mapper anywhere, and provide endpoints to the engine which will be called. This keeps the maintenance of the mapper on the mapper developer side, but increases the latency of mapper.
+![Self-hosted Standalone Mapper](./assets/008/Mapper-2.png)
+
+In the 2nd way, we should post the AWS credentials to the mapper through the internet, which should be done in a secure way. In both ways I don't see another possible approach than posting the AWS credentials to the mapper, since it needs them to query AWS APIs. Anyway, we should warn people to not use a mapper from untrusted authors, because the credentials will be shared with them.
+
+A positive point for having standalone mappers is that it's language agnostic: mapper developers can write the logic in whatever language they want, and it's fine, until they follow the guidelines for the Engine -> Mapper communication protocol.
+
+One negative point is that they'll be provided with the AWS credentials, and they can perform out-of-band data extraction on it on both of the above ways.
```

Comment:
This is a pretty massive negative. With the other approaches the 3rd party module developers are constrained to the APIs we give them so exfiltrating secrets should be much more difficult, and potentially even a tractable problem we can lock down, while these two are basically the wild west -- end-users have to trust us *and* everyone who contributes code to the ecosystem? That's a much taller order.

Reply:
One other thing this doesn't cover is: *all* of the modules have some portion that runs within our engine and within Postgres -- they have to somehow transmit the relevant schema install/uninstall logic to define entities and the engine has to register a mapper object related to it and then funnel the DB/Cloud CRUD operations (and equality, etc) to these external mappers.

*That* means this entire approach, beyond being a worse security situation, is *also* much harder for us to build/maintain. This is also something worth calling out in the RFC.

Reply:
I think it is worth explicitly calling out in the RFC that the security profile of this approach is *much* worse than other approaches because you can't inspect the source code to make sure it's not doing anything malicious, so any responsible developer would have to assume malicious intent.